napari/_qt/_qplugins/_qnpe2.py

```diff
@@ -0,0 +1,232 @@
+                    Viewer,
+                ):
+                    widget_param = param.name
+                    break
+
+    # For magicgui type widget contributions, `Viewer` injection is done by
+    # `magicgui.register_type`.
+    elif isinstance(widget_callable, MagicFactory) or inspect.isfunction(
+        widget_callable
+    ):
+        widget_param = ""
+    else:
+        raise TypeError(
+            trans._(
+                "'{widget}' must be `QtWidgets.QWidget` or `magicgui.widgets.Widget` subclass, `MagicFactory` instance or function. Please raise an issue in napari GitHub with the plugin and widget you were trying to use.",
+                deferred=True,
+                widget=widget_name,
+            )
+        )
+    return widget_param
+
+
+def _toggle_or_get_widget(
+    plugin: str,
+    widget_name: str,
+    name: str,
+) -> Optional[Tuple[Union[FunctionGui, QWidget, Widget], str]]:
+    """Toggle if widget already built otherwise return widget.
+
+    Returned widget will be added to main window by a processor.
+    Note for magicgui type widget contributions, `Viewer` injection is done by
+    `magicgui.register_type` instead of a provider via annnotation.
+    """
+    viewer = _provide_viewer()
+    if viewer is None:
+        raise RuntimeError(
+            trans._(
+                "No current `Viewer` found. Note that widgets cannot be opened in headless mode.",
+                deferred=True,
+            )
+        )
+
+    window = viewer.window
+    if window and (dock_widget := window._dock_widgets.get(name)):
+        dock_widget.setVisible(not dock_widget.isVisible())
+        return None
+
+    # Get widget param name (if any) and check type
+    if widget_contribution := get_widget_contribution(
+        plugin,
+        widget_name,
+    ):
+        widget_callable, _ = widget_contribution
+        widget_param = _get_widget_viewer_param(widget_callable, widget_name)
+
+        kwargs = {}
+        if widget_param:
+            kwargs[widget_param] = viewer
+        return widget_callable(**kwargs), name
+    raise RuntimeError(
+        trans._(
+            "{widget} from {plugin} was not found. Check widget implemented correctly by plugin.",
+            deferred=True,
+            widget=widget_name,
+            plugin=plugin,
+        )
+    )
+
+
+def _get_current_dock_status(name: str) -> bool:
+    window = _provide_window()
+    if window is None:
+        raise RuntimeError(
+            trans._(
+                "No current `Window` found. Note that widgets cannot be opened in headless mode.",
+                deferred=True,
+            )
+        )
+    if name in window._dock_widgets:
+        return window._dock_widgets[name].isVisible()
+    return False
+
+
+def _get_widgets_submenu_actions(
+    mf: PluginManifest,
+) -> Tuple[List[Tuple[str, SubmenuItem]], List[Action]]:
+    """Get widget submenu and actions for a single npe2 plugin manifest."""
```

Comment:
```suggestion
def _build_widgets_submenu_actions(
    mf: PluginManifest,
) -> Tuple[List[Tuple[str, SubmenuItem]], List[Action]]:
    """Build widget submenu and actions for a single npe2 plugin manifest."""
```

For me `get` is getting something already existing, this functions build the submenu